Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -530,69 +530,198 @@ const char* TerrainSystem::GetMaxSurfaceName(
     return "";
 }
 
-/*
+void TerrainSystem::ProcessHeightsFromList(
+    const AZStd::vector<AZ::Vector3>& inPositions,
+    AzFramework::Terrain::HeightListFillCallback perPositionCallback,
+    Sampler sampleFilter) const
+{
+    if (!perPositionCallback)
+    {
+        return;
+    }
+
+    for (const auto& position : inPositions)
+    {
+        bool terrainExists = false;
+        float height = GetHeight(position, sampleFilter, &terrainExists);
+        perPositionCallback(position, height, terrainExists);
+    }
+}
+
+void TerrainSystem::ProcessNormalsFromList(
+    const AZStd::vector<AZ::Vector3>& inPositions,
+    AzFramework::Terrain::NormalListFillCallback perPositionCallback,
+    Sampler sampleFilter) const
+{
+    if (!perPositionCallback)
+    {
+        return;
+    }
+
+    for (const auto& position : inPositions)
+    {
+        bool terrainExists = false;
+        AZ::Vector3 normal = GetNormal(position, sampleFilter, &terrainExists);
+        perPositionCallback(position, normal, terrainExists);
+    }
+}
+
+void TerrainSystem::ProcessSurfaceWeightsFromList(
+    const AZStd::vector<AZ::Vector3>& inPositions,
+    AzFramework::Terrain::SurfaceWeightsListFillCallback perPositionCallback,
+    Sampler sampleFilter) const
+{
+    if (!perPositionCallback)
+    {
+        return;
+    }
+
+    for (const auto& position : inPositions)
+    {
+        bool terrainExists = false;
+        AzFramework::SurfaceData::SurfaceTagWeightList surfaceWeights;
```

Comment:
You should be able to just construct this once outside the for loop.

Reply:
Fixed